### PR TITLE
[translation] Deprecated DiffOperation

### PR DESCRIPTION
**Pull request by zerustech**
## Summary:

The `DiffOperation` class has been deprecated and `TargetOperation`
should be used instead, because `DiffOperation` has nothing to do
with 'diff', thus its class name is misleading.

Also added detailed documents for all operation interface and classes.
## Background:

The following names should have consistent meanings for all operations:

The name of `intersection` is temporarily introduced here to explain this issue.
- [x] `intersection` = source &#x2229; target = {x: x &#x2208; source &#x2227; x &#x2208; target}
- [x] `all` = **result of the operation, depends on the operation.**
- [x] `new` = all &#x2216; source = {x: x &#x2208; all &#x2227; x &#x2209; source}
- [x] `obsolete` = source &#x2216; all = {x: x &#x2208; source &#x2227; x &#x2209; all}

The following analysis explains why `DiffOperation` should be deprecated.
## Logic of `MergeOperation`:
- [x] `all` = source &#x222a; target = {x: x &#x2208; source &#x2228; x &#x2208; target}
- [x] `new` = all &#x2216; source = {x: x &#x2208; target &#x2227; &#x2209; source}
- [x] `obsolete` = source &#x2216; all = {x: x &#x2208; source &#x2227; x &#x2209; source &#x2227; x &#x2209; target} = &#x2205;

This absolutely makes sense.
## Logic of `DiffOperation`:
- [ ] `all` =  intersection &#x222a; (target &#x2216; intersection) = target
- [x] `new` = all &#x2216; source = {x: x &#x2208; target &#x2227; x &#x2209; source}
- [x] `obsolete` = source &#x2216; all = source &#x2216; target = {x: x &#x2208; source &#x2227; x &#x2209; target}

The `all` part is confusing because 'diff' should either mean 'relative complement' or 'symmetric difference' operation:
### Relative Complement:
- `all` = source &#x2216; target = {x: x &#x2208; source &#x2227; x &#x2209; target}
### Symmetric Difference:
- `all` = (source &#x2216; target) &#x222a; (target &#x2216; source) = {x: x &#x2208; source &#x2227; x &#x2209; target &#x2228; x &#x2208; target &#x2227; x &#x2209; source}
### Current Logic has Nothing to do with "Diff":
- `all` =  intersection &#x222a; (target &#x2216; intersection) = target

So the name of `DiffOperation` is misleading and inappropriate.
Unfortunately, there is no corresponding set operation for this class,
so it's hard to give it an apppriate name.
From my point of view, I believe the most accurate name for this class
should be `TargetOperation` because its result is same as the target set.

| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | yes |
| Tests pass? | yes |
| Fixed tickets | n/a |
| License | MIT |
| Doc PR | n/a |

### CODE REVIEW

**Comment by zerustech** [comment]
@stof @aitboudad Please check, thanks.

File: src/Symfony/Component/Translation/Catalogue/TargetOperation.php [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+/**
+ * Target operation between two catalogues:
+ * intersection = source ∩ target = {x: x ∈ source ∧ x ∈ target}
+ * all = intersection ∪ (target ∖ intersection) = target
+ * new = all ∖ source = {x: x ∈ target ∧ x ∉ source}
+ * obsolete = source ∖ all = source ∖ target = {x: x ∈ source ∧ x ∉ target}
+ * Basically, the result contains messages from the target catalogue. 
+ *
+ * @author Jean-François Simon <contact@jfsimon.fr>
```
Comment (OskarStark) [inline src/Symfony/Component/Translation/Catalogue/TargetOperation.php]: if you create this class and it is new, is it correct to copy the author?

Comment (WouterJ) [inline src/Symfony/Component/Translation/Catalogue/TargetOperation.php]: YEs, author indicates who created the code. This way, you know who to ask for if you're really stuck with a bit of code. That's why I propose to remvoe some of the added `@author` tags (for instance, from the deprecated class), as the current PR doesn't add much to it.

Comment (zerustech) [inline src/Symfony/Component/Translation/Catalogue/TargetOperation.php]: @OskarStark I copied the author tags because I didn't change the code at all, I just copied the original code from the deprecated class.

Comment (zerustech) [inline src/Symfony/Component/Translation/Catalogue/TargetOperation.php]: @WouterJ I added myself as an additional author because I added detailed comments for these classes. They were a bit confusing before and it takes time to figure out the correct meanings of the classes as well as their properties.

Comment (stof) [inline src/Symfony/Component/Translation/Catalogue/TargetOperation.php]: @zerustech ``@author`` is mostly useful for the author of the code, as explained by @WouterJ. Adding only comments does not really need to be added as ``@author`` (these annotations are not a way to track all contributors to the class. This is what the git history is for)

Comment (zerustech) [inline src/Symfony/Component/Translation/Catalogue/TargetOperation.php]: @stof I've updated the ``@author`` tags

**Comment by aitboudad** [comment]
could you update https://github.com/symfony/symfony/blob/2.8/src/Symfony/Bundle/FrameworkBundle/Command/TranslationUpdateCommand.php#L142 too

**Comment by zerustech** [comment]
@aitboudad Shall I create a separate PR for it after this PR is merged or just include the change in this PR?

**Comment by aitboudad** [comment]
just include the change

**Comment by zerustech** [comment]
@aitboudad Updated

**Comment by aitboudad** [comment]
:+1:

**Comment by aitboudad** [comment]
ping @symfony/deciders

File: src/Symfony/Component/Translation/Tests/Catalogue/DiffOperationTest.php [hunk 2]
```diff
@@ -15,6 +15,9 @@
 use Symfony\Component\Translation\MessageCatalogue;
 use Symfony\Component\Translation\MessageCatalogueInterface;
 
+/**
+ * @group legacy
+ */
 class DiffOperationTest extends AbstractOperationTest
```
Comment (xabbuh) [inline src/Symfony/Component/Translation/Tests/Catalogue/DiffOperationTest.php]: What about extending the `TargetOperationTest` class instead and overriding only the `createOperation` method instead?

File: src/Symfony/Component/Translation/Catalogue/AbstractOperation.php [hunk 3]
```diff
@@ -17,38 +17,69 @@
...
+ * source (the left argument) and target (the right argument) catalogues and
+ * stores the following results:
+ *
+ * Messages: also called 'all', are valid messages for the given domain after the operation is performed.
+ *
+ * New Messages: also called 'new' (new = all ∖ source = {x: x ∈ all ∧ x ∉ source}).
+ *
+ * Obsolete Messages: also called 'obsolete' (obsolete = source ∖ all = {x: x ∈ source ∧ x ∉ all}).
+ *
+ * Result: also called 'result', is the resulting catalogue for the given domain that holds the same messages as 'all'.
```
Comment (stof) [inline src/Symfony/Component/Translation/Catalogue/AbstractOperation.php]: this is already documented in the interface. No need to duplicate it here (duplication of documentation makes maintenance harder as we have to remember to update it everywhere)

**Comment by zerustech** [comment]
@xabbuh @stof Updated

**Comment by xabbuh** [comment]
Looks like you need to bump the required version of the Translation component in the FrameworkBundle.

**Comment by zerustech** [comment]
@xabbuh Updated

File: src/Symfony/Bundle/FrameworkBundle/Tests/Command/TranslationUpdateCommandTest.php [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+ */
+
+namespace Symfony\Bundle\FrameworkBundle\Tests\Command;
+
+use Symfony\Component\Console\Application;
+use Symfony\Component\Console\Tester\CommandTester;
+use Symfony\Bundle\FrameworkBundle\Command\TranslationUpdateCommand;
+use Symfony\Component\Filesystem\Filesystem;
+
+class TranslationUpdateCommandTest extends \PHPUnit_Framework_TestCase
```
Comment (aitboudad) [inline src/Symfony/Bundle/FrameworkBundle/Tests/Command/TranslationUpdateCommandTest.php]: I think this should be added in `2.3`

Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/Tests/Command/TranslationUpdateCommandTest.php]: Testing the command is indeed a different topic and should go in a separate PR in 2.3

File: src/Symfony/Component/Translation/Tests/Catalogue/TargetOperationTest.php [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Translation\Tests\Catalogue;
+
+use Symfony\Component\Translation\Catalogue\TargetOperation;
+use Symfony\Component\Translation\MessageCatalogueInterface;
+
+class TargetOperationTest extends DiffOperationTest
```
Comment (xabbuh) [inline src/Symfony/Component/Translation/Tests/Catalogue/TargetOperationTest.php]: Inheritance should be the other way around. Otherwise, the new `TargetOperationTest` is a legacy test too.

Comment (zerustech) [inline src/Symfony/Component/Translation/Tests/Catalogue/TargetOperationTest.php]: @xabbuh Fixed

File: src/Symfony/Bundle/FrameworkBundle/Tests/Command/TranslationUpdateCommandTest.php (hunk 4 above)
Comment (zerustech) [inline src/Symfony/Bundle/FrameworkBundle/Tests/Command/TranslationUpdateCommandTest.php]: @aitboudad @stof I've created PR #15673 as per your suggestion. However, if test script is added in a separate PR, how to prove the issue (and avoiding regressions) in this PR? And why should it be added in ``2.3``?

Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/Tests/Command/TranslationUpdateCommandTest.php]: @zerustech because testing the command makes sense in all versions of Symfony, and so it should be added first in 2.3 (which will then be merged into newer branches)

Comment (zerustech) [inline src/Symfony/Bundle/FrameworkBundle/Tests/Command/TranslationUpdateCommandTest.php]: @stof I agree with you, but why other test scripts like ``TranslationDebugCommandTest.php`` was not added in ``2.3``?

Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/Tests/Command/TranslationUpdateCommandTest.php]: because the TranslationDebugCommand does not exist in 2.3, and so cannot be tested there...

Comment (zerustech) [inline src/Symfony/Bundle/FrameworkBundle/Tests/Command/TranslationUpdateCommandTest.php]: @stof So just to clarify the workflow for adding a new test script:
* Find the oldest supported branch that contains the file to be tested.
    In this case, the branch is ``2.3``, but in the future if ``2.3`` is not supported any more, 
    the branch might be ``2.7`` for example.

*  Add the test script in that branch

Is that correct?

Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/Tests/Command/TranslationUpdateCommandTest.php]: yes

Comment (zerustech) [inline src/Symfony/Bundle/FrameworkBundle/Tests/Command/TranslationUpdateCommandTest.php]: @stof Thanks, I will follow it in the future.

**Comment by xabbuh** [comment]
:+1:

**Comment by aitboudad** [comment]
Thank you @zerustech.

**Comment by zerustech** [comment]
@aitboudad No problem.
